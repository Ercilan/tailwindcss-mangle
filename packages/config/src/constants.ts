export const CONFIG_NAME = 'tailwindcss-mangle'
