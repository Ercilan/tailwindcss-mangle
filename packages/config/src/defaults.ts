import process from 'node:process'
import { defaultMangleClassFilter } from '@tailwindcss-mangle/shared'
import { CSS_LANGS_RE } from 'is-css-request'
import type { MangleUserConfig, PatchUserConfig, UserConfig } from './types'

const defaultPipelineInclude = ['**/*.{html,js,ts,jsx,tsx,vue,svelte,astro,elm,php,phtml,mdx,md}', CSS_LANGS_RE]

const defaultPipelineExclude = [/[\\/](node_modules|dist|\.temp|\.cache|\.vscode)[\\/]/]

export function getDefaultPatchConfig(): PatchUserConfig {
  return {
    output: {
      filename: '.tw-patch/tw-class-list.json',
      removeUniversalSelector: true,
      loose: true,
    },
    tailwindcss: {},
  }
}

export function getDefaultMangleUserConfig(): MangleUserConfig {
  return {
    mangleClassFilter: defaultMangleClassFilter,
    include: defaultPipelineInclude,
    exclude: defaultPipelineExclude,
    disabled: process.env.NODE_ENV === 'development',
    classListPath: '.tw-patch/tw-class-list.json',
    classMapOutput: {
      enable: false,
      filename: '.tw-patch/tw-map-list.json',
      loose: true,
    },
    preserveFunction: [],
  }
}

export function getDefaultUserConfig(): UserConfig {
  return {
    patch: getDefaultPatchConfig(),
    mangle: getDefaultMangleUserConfig(),
  }
}
