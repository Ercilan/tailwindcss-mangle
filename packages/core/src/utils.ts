import micromatch from 'micromatch'

const { isMatch } = micromatch

export function createGlobMatcher(pattern: string | string[] | undefined, fallbackValue: boolean = false) {
  if (pattern === undefined) {
    return function () {
      return fallbackValue
    }
  }
  return function (file: string) {
    return isMatch(file, pattern)
  }
}

export { defaultMangleClassFilter, isMap, isRegexp } from '@tailwindcss-mangle/shared'
