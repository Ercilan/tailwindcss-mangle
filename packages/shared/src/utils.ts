import type { IClassGeneratorOptions, IClassGenerator } from './types'

export const defaultMangleClassFilter = (className: string) => {
  // ignore className like 'filter','container'
  // it may be dangerous to mangle/rename all StringLiteral , so use /-/ test for only those with /-/ like:
  // bg-[#123456] w-1 etc...
  return /[:-]/.test(className)
}

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new TypeError('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new TypeError('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (const item of arr) {
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (Array.isArray(bucket)) {
      result[bucketCategory].push(item)
    } else {
      result[bucketCategory] = [item]
    }
  }

  return result
}

export const acceptChars = [...'abcdefghijklmnopqrstuvwxyz']

export function stripEscapeSequence(words: string) {
  return words.replaceAll('\\', '')
}

export const validate = (opts: IClassGeneratorOptions, classGenerator: IClassGenerator) => {
  if (!opts.log) return
  for (const className in classGenerator.newClassMap) {
    const c = classGenerator.newClassMap[className]
    if (c.usedBy.length > 0) {
      continue
    }
    if (/.+\.css:*$/.test(c.usedBy[0])) {
      console.log(`The class name '${className}' is not used: defined at ${c.usedBy[0]}.`)
    } else {
      console.log(`The class name '${className}' is not defined: used at ${c.usedBy[0]}.`)
    }
  }
}

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}

export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (typeof item === 'string') {
        if (item === str) {
          return true
        }
      } else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError("paramater 'arr' should be a Array of Regexp | String !")
}
