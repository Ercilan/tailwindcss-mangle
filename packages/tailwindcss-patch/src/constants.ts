export const pkgName = 'tailwindcss-patch'
