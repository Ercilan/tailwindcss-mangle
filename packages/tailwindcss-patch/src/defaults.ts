import type { DeepRequired, InternalPatchOptions, PatchOptions } from './types'
import process from 'node:process'
import { defu } from '@tailwindcss-mangle/shared'

export function getDefaultPatchOptions(): DeepRequired<PatchOptions> {
  return {
    applyPatches: {
      exportContext: true,
      extendLengthUnits: false,
    },
    overwrite: true,
  }
}

export function getPatchOptions(options?: PatchOptions) {
  return defu<InternalPatchOptions, Partial<InternalPatchOptions>[]>(
    options,
    {
      basedir: process.cwd(),
    },
    getDefaultPatchOptions(),
  )
}
