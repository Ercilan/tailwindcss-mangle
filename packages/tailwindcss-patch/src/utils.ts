import fss from 'node:fs'
import fs from 'node:fs/promises'
import path from 'node:path'
import type { SyncOpts } from 'resolve'
import pkg from 'resolve'
import type { PackageJson } from 'pkg-types'

const { sync } = pkg

export function ensureFileContent(filepaths: string | string[]) {
  if (typeof filepaths === 'string') {
    filepaths = [filepaths]
  }
  let content
  for (const filepath of filepaths) {
    if (fss.existsSync(filepath)) {
      content = fss.readFileSync(filepath, {
        encoding: 'utf8',
      })
      break
    }
  }
  return content
}

export function requireResolve(id: string, opts?: SyncOpts) {
  return sync(id, opts)
}

export async function ensureDir(p: string) {
  try {
    await fs.access(p)
  }
  catch {
    await fs.mkdir(p, {
      recursive: true,
    })
  }
}

function searchPackageJSON(dir: string) {
  let packageJsonPath
  while (true) {
    if (!dir) {
      return
    }
    const newDir = path.dirname(dir)
    if (newDir === dir) {
      return
    }
    dir = newDir
    packageJsonPath = path.join(dir, 'package.json')
    if (fss.existsSync(packageJsonPath)) {
      break
    }
  }

  return packageJsonPath
}

function getTailwindcssEntry(name: string = 'tailwindcss', opts?: SyncOpts) {
  return requireResolve(name, opts)
}

function getPackageJsonPath(name: string, options: SyncOpts = {}) {
  const entry = getTailwindcssEntry(name, options)
  if (!entry) {
    return
  }

  return searchPackageJSON(entry)
}

export function getPackageInfoSync(name: string, options: SyncOpts = {}) {
  const packageJsonPath = getPackageJsonPath(name, options)
  if (!packageJsonPath) {
    return
  }

  const packageJson: PackageJson = JSON.parse(fss.readFileSync(packageJsonPath, 'utf8'))

  return {
    name,
    version: packageJson.version,
    rootPath: path.dirname(packageJsonPath),
    packageJsonPath,
    packageJson,
  }
}

export function isObject(val: any) {
  return val != null && typeof val === 'object' && Array.isArray(val) === false
};
