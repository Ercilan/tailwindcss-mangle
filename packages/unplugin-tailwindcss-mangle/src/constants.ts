export const pluginName = 'unplugin-tailwindcss-mangle'

export const webpackCacheKey = pluginName + '-sources-cache-key'
