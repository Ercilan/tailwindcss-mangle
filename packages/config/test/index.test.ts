import { deleteAsync } from 'del'
import { existsSync } from 'fs-extra'
import { resolve } from 'pathe'
import { CONFIG_NAME } from '@/constants'
import { getDefaultTransformerConfig, getDefaultUserConfig } from '@/defaults'
import { getConfig, initConfig } from '@/index'
import { fixturesRoot } from './utils'

function normaliseRegex(value: any): any {
  if (Array.isArray(value)) {
    return value.map(normaliseRegex)
  }
  if (value instanceof RegExp) {
    return value.toString()
  }
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value).map(([key, val]) => [key, normaliseRegex(val)]))
  }
  return value
}

describe('config', () => {
  it('0.default', async () => {
    const cwd = resolve(fixturesRoot, './config/0.default')
    const configPath = resolve(cwd, `${CONFIG_NAME}.config.ts`)
    if (existsSync(configPath)) {
      await deleteAsync(configPath)
    }

    await initConfig(cwd)
    expect(existsSync(configPath)).toBe(true)

    const { config } = await getConfig(cwd)
    expect(normaliseRegex(config)).toEqual(normaliseRegex(getDefaultUserConfig()))
  })

  it('1.change-options', async () => {
    const cwd = resolve(fixturesRoot, './config/1.change-options')
    const { config } = await getConfig(cwd)
    expect(config).toEqual({
      registry: {
        output: {
          file: 'xxx/yyy/zzz.json',
          pretty: false,
          stripUniversalSelector: false,
        },
        tailwind: {
          cwd: 'aaa/bbb/cc',
        },
      },
      transformer: getDefaultTransformerConfig(),
    })
  })

  it('2.transformer-options', async () => {
    const cwd = resolve(fixturesRoot, './config/2.transformer-options')
    const { config } = await getConfig(cwd)
    expect(normaliseRegex(config)).toMatchSnapshot()
  })
})
