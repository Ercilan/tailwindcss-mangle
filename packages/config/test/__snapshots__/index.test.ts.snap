// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > 2.transformer-options 1`] = `
{
  "registry": {
    "output": {
      "file": "xxx/yyy/zzz.json",
      "pretty": false,
      "stripUniversalSelector": false,
    },
    "tailwind": {
      "cwd": "aaa/bbb/cc",
    },
  },
  "transformer": {
    "disabled": false,
    "filter": [Function],
    "generator": {
      "log": true,
    },
    "preserve": {
      "classes": [],
      "functions": [],
    },
    "registry": {
      "file": "zzzzz.json",
      "mapping": {
        "enabled": true,
        "file": "ffff.json",
        "loose": false,
      },
    },
    "sources": {
      "exclude": [],
      "include": [
        "/\\.(html|js|ts|jsx|tsx|vue|svelte|astro|elm|php|phtml|mdx|md)(?:$|\\?)/",
        "/\\.(css|less|sass|scss|styl|stylus|pcss|postcss|sss)(?:$|\\?)/",
      ],
    },
  },
}
`;
